### ISSUE TITLE
Proposal: merge packs

### ISSUE THREAD

**Issue by zekenie** [rubyatscale/packs#143]
Hey!

We have a fairly large modularization effort going on and all the tools you all have created have been very very helpful. Thank you.

I found something I think I want `bin/packs` to do: merge packs together. We've overgranularized a bit and we've identified some packs that we want smooshed together. Of course it's not terrifically hard to do by hand... but it'd be nice to be able to do it with just a few key strokes.

```
bin/packs merge packs/auth packs/permissions
```

This would take files from `permissions` and put them into `auth`. If there were any name conflicts (which there shouldn't be in most cases for Zeitwerk to work, i think?) it could fail with a helpful error message. Any dependencies that `permissions` had that `auth` didn't have would get copied into the `auth/package.yml` file.

If you like the idea I might try to add it in a PR. Then again I might be being too optimistic about my time.

**Comment by ashleywillard** [comment]
Hi there! We think that this sounds like a good idea. There are a few challenges that one might face when implementing this functionality:

1. What if the settings in the `package.yml` files differ? For example, if `packs/auth` has `enforce_dependencies` set to `true` but `packs/permissions` has it set to `false`.
2. Namespacing. Let's say we have something like the following. Then we'll want to nest the files directly under auth and change the class constant name.
```
packs
  auth
    app
      services
        auth
  permissions
    app
      services
        permissions
          xyz.rb (class Permissions::Xyz)
```

**Comment by zekenie** [comment]
These are great questions

I hadn't considered your first point about `package.yml` conflicts. I'm not sure. I see a few options--none of which i'm sold on:

1. Do something like what git does and make the user manually reconcile 
2. Do something like `Object.assign` in js where there's a clear winner if anything differs
3. Some version of 2 where you surface what happened with a warning
4. Some sort of failure that says "You can't merge these packs until their package yml files are compatible. `packs/permissions` has `enforce_dependencies` set to `false` but `packs/authentication` has it sent to true.

On your second point, I think I have an opinion. I don't think this tool would change any namespacing. I think it would be intended to recursively merge the file trees and leave the ruby files untouched similar to how `bin/packs move` works. I'm personally working with files that haven't been well namespaced yet, so for me this is more about chunking. I kind of see getting the name spaces as a separate tool. I might be missing something though

**Comment by professor** [comment]
There's a `bin/packs` command that will move a file or a directory and I'm curious if that will work for you in this case.

```
packs move packs/destination_pack path/to/file.rb path/to/directory
```

**Comment by zekenie** [comment]
@professor, good question. Let's say I had two packs, `orders` and `carts` and I decide i really want them to be the same pack. I don't want cart to exist, i just want orders, and i want the contents of cart to be recursively merged into orders. One models directory, one controllers directory, etc. I don't think `packs move` can do that, but i might be wrong. I think I could move `cart` _into_ `orders`, but not merge them

**Comment by professor** [comment]
I tried it with `packs move` and the relative paths worked fine. You should end up with a relatively empty `carts` pack that would be easy to delete. (e.g. remove the package.yml and folder)

**Comment by zekenie** [comment]
What was the command you ran to make that work? Something like:

```
bin/packs move packs/orders packs/carts
```

?

I have to say i never tried that! if this works than perhaps I just didn't understand what `move` could do!

**Comment by ashleywillard** [comment]
Hi @zekenie! Did this approach work for you?

**Comment by zekenie** [comment]
Yes! it looks like it does. Sorry, I forgot about this issue. I do think the semantics of `merge` are better when you're talking about two packs rather than `move` which makes more sense for files, but that's a rather minor point. Thank you for building this!
